beacon-chain/sync/initial-sync/blocks_fetcher_utils.go

```diff
@@ -119,6 +141,139 @@ func (f *blocksFetcher) nonSkippedSlotAfter(ctx context.Context, slot uint64) (u
+	}
+
+	// The current slot's epoch must be after the finalization epoch,
+	// triggering backtracking on earlier epochs is unnecessary.
+	finalizedEpoch := f.chain.FinalizedCheckpt().Epoch
+	epoch := helpers.SlotToEpoch(slot)
+	if epoch <= finalizedEpoch {
+		return nil, errors.New("slot is not after the finalized epoch, no backtracking is necessary")
+	}
+	// Update slot to the beginning of the current epoch (preserve original slot for comparison).
+	slot, err := helpers.StartSlot(epoch)
+	if err != nil {
+		return nil, err
+	}
+
+	// Select peers that have higher head slot, and potentially blocks from more favourable fork.
+	// Exit early if no peers are ready.
+	_, peers := f.p2p.Peers().BestNonFinalized(1, epoch+1)
+	if len(peers) == 0 {
+		return nil, errNoPeersAvailable
+	}
+	f.rand.Shuffle(len(peers), func(i, j int) {
+		peers[i], peers[j] = peers[j], peers[i]
+	})
+
+	// Query all found peers, stop on peer with alternative blocks, and try backtracking.
+	for i, pid := range peers {
+		log.WithFields(logrus.Fields{
+			"peer": pid,
+			"step": fmt.Sprintf("%d/%d", i+1, len(peers)),
+		}).Debug("Searching for alternative blocks")
+		fork, err := f.findForkWithPeer(ctx, pid, slot)
+		if err != nil {
+			log.WithFields(logrus.Fields{
+				"peer":  pid,
+				"error": err.Error(),
+			}).Debug("No alternative blocks found for peer")
+			continue
+		}
+		return fork, nil
+	}
+	return nil, errNoPeersWithAltBlocks
+}
+
+// findForkWithPeer loads some blocks from a peer in an attempt to find alternative blocks.
+func (f *blocksFetcher) findForkWithPeer(ctx context.Context, pid peer.ID, slot uint64) (*forkData, error) {
+	// Locate non-skipped slot, supported by a given peer (can survive long periods of empty slots).
+	// When searching for non-empty slot, start an epoch earlier - for those blocks we
+	// definitely have roots. So, spotting a fork will be easier. It is not a problem if unknown
+	// block of the current fork is found: we are searching for forks when FSMs are stuck, so
+	// being able to progress on any fork is good.
+	pidState, err := f.p2p.Peers().ChainState(pid)
+	if err != nil {
+		return nil, fmt.Errorf("cannot obtain peer's status: %w", err)
+	}
+	slotsPerEpoch := params.BeaconConfig().SlotsPerEpoch
+	nonSkippedSlot, err := f.nonSkippedSlotAfterWithPeersTarget(
+		ctx, slot-slotsPerEpoch, []peer.ID{pid}, helpers.SlotToEpoch(pidState.HeadSlot))
+	if err != nil {
+		return nil, fmt.Errorf("cannot locate non-empty slot for a peer: %w", err)
+	}
+
+	// Request blocks starting from the first non-empty slot.
+	req := &p2ppb.BeaconBlocksByRangeRequest{
+		StartSlot: nonSkippedSlot,
+		Count:     slotsPerEpoch * 2,
+		Step:      1,
+	}
+	blocks, err := f.requestBlocks(ctx, req, pid)
+	if err != nil {
+		return nil, fmt.Errorf("cannot fetch blocks: %w", err)
+	}
+
+	// Traverse blocks, and if we've got one that doesn't have parent in DB, backtrack on it.
+	for _, block := range blocks {
```

Comment:
Doing this for all blocks might be overkill, we ensure that the blocks that we receive are `linear` so we just need to check for the first one.

Reply:
We are looking for block that is not in our cache/db, and that is not guaranteed to be the first one (we can't know where diverging block will appear). If it is, no problem either - loop will terminate immediately. 

We do not know the exact point where forks diverge, hence looking through a window of blocks for each potential peer (peer advertising high enough head slot). It is assumed, that alternative forks are more favourable and numerous (most of the times), so unknown block is to be found quickly: one of the first blocks pulled, really.

Reply:
Unless I am missing something don't we validate and ensure that range requests with a step of `1` return only linear blocks ? So each block in the list will have its parent in the previous index. Once this has been validated, we only need to ensure that the first block's parent is in our db as we know that the corresponding blocks are all from the same branch.